modules/integration/tests-common/jacoco-report-generator/src/main/java/org/wso2/carbon/identity/jacoco/ReportGenerator.java

```diff
@@ -0,0 +1,198 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.jacoco;
+
+import org.codehaus.plexus.util.FileUtils;
+import org.jacoco.core.analysis.Analyzer;
+import org.jacoco.core.analysis.CoverageBuilder;
+import org.jacoco.core.analysis.IBundleCoverage;
+import org.jacoco.core.tools.ExecFileLoader;
+import org.jacoco.report.DirectorySourceFileLocator;
+import org.jacoco.report.IReportVisitor;
+import org.jacoco.report.xml.XMLFormatter;
+
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * This will create an XML report projects based on a
+ * single execution data store called jacoco.exec.
+ */
+public class ReportGenerator {
+
+    public static final String CLASS_FILE_PATTERN = "**/*.class";
+
+    private final String title;
+    private final File executionDataFile;
+    private final Set<File> classDirectories;
+    private final File xmlReport;
+    private final File tempDirectory;
+
+    private ExecFileLoader execFileLoader;
+
+    /**
+     * Create a new generator based for the given project.
+     *
+     * @param executionDataFile the execution data file
+     * @param classDirectories  the set of class directories
+     */
+    public ReportGenerator(File executionDataFile, Set<File> classDirectories) {
+
+        this.title = "Jacoco Coverage Report";
+        this.executionDataFile = executionDataFile;
+        this.classDirectories = classDirectories;
+        this.xmlReport = new File("./report/jacoco.xml");
+        this.tempDirectory = new File("./tmp");
+
+        // Create report directory if it does not exist
+        File reportDir = this.xmlReport.getParentFile();
+        if (!reportDir.exists() && !reportDir.mkdirs()) {
+            throw new RuntimeException("Failed to create report directory: " + reportDir.getAbsolutePath());
+        }
+    }
+
+    /**
+     * Create the report.
+     *
+     * @throws IOException
+     */
+    public void create() throws IOException {
+
+        // Read the jacoco.exec file. Multiple data files could be merged
+        // at this point
+        loadExecutionData();
+
+        // Run the structure analyzer on a single class folder to build up
+        // the coverage model. The process would be similar if your classes
+        // were in a jar file. Typically you would create a bundle for each
+        // class folder and each jar you want in your report. If you have
+        // more than one bundle you will need to add a grouping node to your
+        // report
+        final IBundleCoverage bundleCoverage = analyzeStructure();
+
+        createReport(bundleCoverage);
+    }
+
+    private void createReport(final IBundleCoverage bundleCoverage) throws IOException {
+
+        // Create a concrete report visitor based on some supplied
```

Comment:
Move as method comment?

Reply:
Done with https://github.com/wso2/product-is/pull/21108